File: Napakalaki Ruby/SpecificBadConsequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module NapakalakiGame

class SpecificBadConsequence < BadConsequence
  
  def initialize(texto,nivel, tVisible, tHidden)
    super(texto, nivel)
    
    @specificVisibleTreasures = tVisible
    @specificHiddenTreasures = tHidden
    
  end
  
  def getSpecificVisibleTreasures
    return @specificVisibleTreasures
  end
  
  def getSpecificHiddenTreasures
    return @specificHiddenTreasures
  end
  
  def to_s
    cadena = super
    
    cadena += " Especificos tesoros visibles perdidos = "
                for i in  @specificVisibleTreasures
                    cadena += i.to_s() + ", "
                end
    cadena += "Especificos tesoros ocultos perdidos = "
             for j in @specificHiddenTreasures 
                 cadena += j.to_s() + ", "
             end
    
    return cadena
  end
  
  def isEmpty
    vacio = false   
    vacio = true if (@specificVisibleTreasures.size == 0 and @specificHiddenTreasures.size == 0)
    
    return vacio
  end
  
  def substractVisibleTreasure(t)
    
    indice = @specificVisibleTreasures.index(t.getType)
    
    @specificVisibleTreasures.delete_at(indice)
  end
  
  def substractHiddenTreasure(t)
    indice = @specificHiddenTreasures.index(t.getType)
    @specificHiddenTreasures.delete_at(indice)
  end
  
  def adjustToFitTreasureList(v, h)
    copiav = Array.new
    copiah = Array.new
    vAjustado = Array.new
    hAjustado = Array.new

    v.each do |tesoro|
     copiav << (tesoro.getType())
    end    

    h.each do |tesoro|
     copiah << (tesoro.getType())
    end  

    @specificVisibleTreasures.each do |tipo|
     if(copiav.include?(tipo)) 
       vAjustado << (tipo)
       copiav.delete_at(copiav.index(tipo))
     end
    end

    @specificHiddenTreasures.each do |tipo|
     if(copiah.include?(tipo)) 
       hAjustado << (tipo)
       copiah.delete_at(copiah.index(tipo))
     end
    end

    malrollo = SpecificBadConsequence.new(@text,@levels,vAjustado,hAjustado)

    return malrollo
  end
  
end

end

File: Napakalaki Java/NapakalakiGame/SpecificBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;

/**
 *
 * @author Germán
 */
public class SpecificBadConsequence extends BadConsequence {
    private ArrayList<TreasureKind> specificVisibleTreasures = null;
    private ArrayList<TreasureKind> specificHiddenTreasures = null;
    
    public SpecificBadConsequence (String Text, int Levels, ArrayList<TreasureKind> tVisible,
                                 ArrayList<TreasureKind> tHidden) {
        super(Text, Levels);
        specificVisibleTreasures = tVisible;
        specificHiddenTreasures = tHidden;
    }
    
    @Override
    public String toString() {
        String cadena;
        
        cadena = "Levels = " + Integer.toString(levels);
                     
        if (specificVisibleTreasures != null) {
            cadena += " Specific Visible Treasures = ";
            for (TreasureKind specificVisibleTreasure : specificVisibleTreasures) 
                cadena += ", " + specificVisibleTreasure.toString();
        }     
        if(specificHiddenTreasures != null) {
            cadena += " Specific Hidden Treasures = ";
            for (TreasureKind specificHiddenTreasure : specificHiddenTreasures) 
                cadena += ", " + specificHiddenTreasure.toString();
        }
        
        return cadena;
    }
    
    @Override
    public boolean isEmpty(){
        return (specificVisibleTreasures.isEmpty() && specificHiddenTreasures.isEmpty());
    }
    
    public ArrayList<TreasureKind> getSpecificVisibleTreasures() {
     
         return specificVisibleTreasures;
    }
     
    public ArrayList<TreasureKind> getSpecificHiddenTreasures() {
     
        return specificHiddenTreasures;
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t){ 
        specificVisibleTreasures.remove(t.getType());     
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t){
        specificHiddenTreasures.remove(t.getType());
    }
    
    @Override
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, ArrayList<Treasure> h) {
        SpecificBadConsequence malrollo;
        
        ArrayList<TreasureKind> copiav = new ArrayList();
        ArrayList<TreasureKind> copiah = new ArrayList();
        ArrayList<TreasureKind> vAjustado = new ArrayList();
        ArrayList<TreasureKind> hAjustado = new ArrayList();
                     
        for(Treasure tesoro: v){
            copiav.add(tesoro.getType());
        }
                      
        for(Treasure tesoro: h){
            copiah.add(tesoro.getType());
        }
                           
        for (TreasureKind  tipo : specificVisibleTreasures){
            if(copiav.contains(tipo)) {
                vAjustado.add(tipo);
                copiav.remove(tipo);
            }
        }
                    
        for (TreasureKind  tipo : specificHiddenTreasures){
            if(copiah.contains(tipo)){                           
                hAjustado.add(tipo);
                copiah.remove(tipo);
            }
        }
                    
        malrollo = new SpecificBadConsequence(text,levels,vAjustado,hAjustado);
        
        return malrollo;
    }
}
